### DISCUSSION TITLE
Audio-tools: How to write a complete WAV file to SD card - with header?

### DISCUSSION THREAD

**Discussion by Nik-mayak** [pschatzmann/arduino-audio-tools#658] [Q&A] [answered]
Thank you so much Phil for your work!
I can't solve one problem. According to your example from the wiki [https://github.com/pschatzmann/arduino-audio-tools/wiki/Encoding-and-Decoding-of-Audio](url) a file on the SD card is created, but without a header. How can available in
arduino-audio-tools means to create a complete file - with a header?

**Comment by pschatzmann** [comment, marked as answer]
If you write it thru a WAVEncoder the header should get added to it automatically! 
This is also how the Webserver examples are working: the Wav header makes that the browser is recognising the file format. 
The header is added when the __first write__ is called before the audio data is written.

**Comment by Nik-mayak** [comment]
There is no WAVEEncoder.h library in arduino-audio-tools. There is a WAVEncoder() class in the CodecWAV.h library

**Comment by Nik-mayak** [comment]
EncodedAudioStream out_stream(&file, new WAVEncoder()); // encode as wav file

**Reply by pschatzmann** [reply]
Yes that't the one and that you can find in the documentation! 
ps. don't forget to close the file when you are done, or you might not see any updates to the file.

**Comment by Nik-mayak** [comment]
void record_end(bool pinStatus, int pin, void* ref){
  if (recording == true){
    Serial.println("Record end.");
    copier.end();
    file.close();
    numbFile ++;
    recording = false;
  }

**Comment by Nik-mayak** [comment]
"Yes that't the one and that you can find in the documentation!"
Which WAVEEncoder.h to use?

**Comment by pschatzmann** [comment]
It's automatically included so you don't need to include it a second time.
I just yesterday tested the webserver on a RP2040, so it is hard to think that this is not working...

**Reply by pschatzmann** [reply]
Did you rewind to the beginning or delete the file before writing to it ? Maybe you are just appending...

**Comment by Nik-mayak** [comment]
Each time the file is overwritten, but always without a header.

#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioKitHAL.h"
#include <SPI.h>
#include <SD.h>

//const char *file_name = "/rec.wav";
uint16_t sample_rate = 44100;
uint8_t channels = 2;  // The stream will have 2 channels
uint8_t numbFile = 1;
String bodyFileName = "/Recorder_WAV/RecordTest"; 
String extFile = ".wav";
String fileName = "";
//AudioKitStream kit;
AudioKitStream kitStr;
//SdFat SD;
File file;
EncodedAudioStream out_stream(&file, new WAVEncoder()); // encode as wav file
StreamCopy copier(out_stream, kitStr); // copies sound to out
bool recording = false;  // flag to make sure that close is only executed one
bool playing = false; 

void record_start(bool pinStatus, int pin, void* ref){
  Serial.println("Recording...");
  fileName = bodyFileName + String(numbFile) + extFile;// fileName.c_str()
  // open the output file
  file = SD.open(fileName, FILE_WRITE);
  // Configure WAVEncoder
  auto cfg_WAVEncoder = WAVEncoder().defaultConfig();
  cfg_WAVEncoder.sample_rate = 44100;
  cfg_WAVEncoder.bits_per_sample = 16;
  cfg_WAVEncoder.channels = 2;
  WAVEncoder().begin(file, cfg_WAVEncoder);
  copier.setCheckAvailableForWrite(false);
  copier.begin(file, kitStr);  
  recording = true; 
}
void setup(){
  Serial.begin(115200);
  while(!Serial); // wait for serial to be ready
  Serial.println();
  Serial.println("**********************  Start SETUP   **********************");
  while(!Serial); // wait for serial to be ready
  AudioLogger::instance().begin(Serial, AudioLogger::/*Error*//*Debug*/Info/*Warning*/);  

  // Open SD drive
  if (!SD.begin(PIN_AUDIO_KIT_SD_CARD_CS)) {
    Serial.println("SDcard Initialization failed!");
    while (1); // stop
  }
  Serial.println("SDcard Initialization done.");

  // setup input and output
  auto cfg = kitStr.defaultConfig(RX_MODE);
  cfg.sd_active = true;
  cfg.bits_per_sample = 16;
  cfg.sample_rate = 44100;
  cfg.channels = 2;
  cfg.input_device = AUDIO_HAL_ADC_INPUT_LINE1;
  kitStr.begin(cfg);
  kitStr.setVolume(1.0);

  // Configure out stream
  auto cfg_out = out_stream.defaultConfig();
  cfg_out.bits_per_sample = 16;
  cfg_out.channels = 2;
  cfg_out.sample_rate = 44100;
  out_stream.begin(cfg_out);

  // record when key 1 is pressed
  kitStr.audioActions().add(PIN_KEY1, record_start, record_end);
  Serial.println("Press Key 1 to record");
  //kitStr.audioActions().add(PIN_KEY2, play_start);
  //Serial.println("Press Key 2 to play");

}

void loop(){

  if(recording == true){
    // record or play file
    copier.copy(); 
  }
  // Process keys
  kitStr.processActions();

}

**Reply by pschatzmann** [reply]
Where is the logic to rewind ? or delete ?
I also don't see an closing of the file. It would be my understanding that you are just appending to the files if they exist

**Comment by Nik-mayak** [comment]
yes i just add files.
In the example [https://github.com/pschatzmann/arduino-audio-tools/wiki/Encoding-and-Decoding-of-Audio] you do not close the file. How do you know what a header file is?

**Reply by pschatzmann** [reply]
> yes i just add files.

Or append if the file is existing !

**Comment by pschatzmann** [comment]
If you look at the content of the WAV file the header starts with the content "RIFF"
https://docs.fileformat.com/audio/wav/

**Reply by Nik-mayak** [reply]
yeah, I know.
fb ff ff ff 09 00 09 00 03 00 07 00 08 00 09 00
01 00 04 00 0e 00 0d 00 10 00 11 00 fb ff f9 ff
06 00 07 00 fa ff fd ff ff ff ff ff ff ff fc ff
06 00 04 00 01 00 05 00 05 00 07 00 09 00 0f
this is a small piece of the beginning of the recorded file.

**Comment by pschatzmann** [comment]
I think your logic to create a decoder looks completly wrong to me! 
I suggest to create on in global memory and just work with that one!
WAVEncoder().begin(file, cfg_WAVEncoder); has no impact at all because you create a temporary object that immediatly gets distroyed again.
I confirm that you need to call the begin() method on the encoder to signal that you are starting (again), so that a header is generated. Otherwise it is only generated for the first file. This is taken care of by calling the begin method of the EncodedAudioStream. Unfortunatly you don't call this as part of you start recording...

If you set the log level to Info you should see the message "Writing Header" in the log

**Comment by Nik-mayak** [comment]
Thank you very much for your help. Everything worked as expected! The file header is recorded, the sound is great! I
Once again, thank you so much for your great work!

**Comment by pschatzmann** [comment]
I guess the remove fails if the file is still open. In order to overwrite the content you can also just move the file pointer to the beginning with file.lseek(0);

**Comment by Nik-mayak** [comment]
Good day. There was one more question about the operation of the audio player from the SD card. If, after stopping the player, remove the SD card and insert it back (or another card), the player will no longer start. Moreover, the initialization of the SD card (sd.begin (config)) is successful. The problem turned out to be in the AudioSourceSDFAT.h file.
 virtual void begin() override {
    TRACED();
    static bool is_sd_setup = false;
        if (!is_sd_setup) {
        if (!sd.begin(*p_cfg)) {
          LOGE("sd.begin failed");
        return;
      }
      is_sd_setup = true;
      LOGI("sd.begin: %d", is_sd_setup);
    }
    idx.begin(start_path, exension, file_name_pattern);
    idx_pos = 0;
  }
The is_sd_setup variable is declared as static. That is, it can only be changed on the first call to the begin() function. I commented out static and now everything works fine. You can remove the card, insert a new one and continue Play.
 /*static*/ bool is_sd_setup = false;

**Comment by pschatzmann** [comment]
For the time being I do not plan to change this. On my drive the sd.begin has a high rate of failing, so I avoid to call this method.
The best thing would be if you just create a subclass and overwrite the begin method.

**Reply by Nik-mayak** [reply]
OK thanks

**Comment by miroslavpetrov** [comment]
@Nik-mayak can you share the final working code as I am also struggling with this...

**Comment by pschatzmann** [comment]
Are you struggling writing a wav file or with replacing the SD w/o restarting ?

**Reply by miroslavpetrov** [reply]
I am using [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/b2ae828efb0b9b390fbe6acb5e63857bc3ec5424/examples/examples-audiokit/streams-audiokit-sd-audiokit/streams-audiokit-sd-audiokit.ino) as a starting point.

My goal is to write the mic input as a wav file instead of raw file. 

I am confused about how to configure the EncodedAudioStream and the StreamCopy. 

These are my globals:
```
const char *file_name = "/rec.wav";
uint16_t sample_rate = 16000;
uint8_t channels = 1;  // The stream will have 2 channels 
AudioKitStream kit;
File file;   // final output stream
EncodedAudioStream out(&file, new WAVEncoder()); // Access audioFile as a stream
StreamCopy copier(file, out);
bool recording = false;  // flag to make sure that close is only executed one
uint64_t end_time; // trigger to call endRecord
```

and this is how i tried to change the record_start function:

```
void record_start(bool pinStatus, int pin, void* ref){
  Serial.println("Recording...");
  // open the output file
  file = SD.open(file_name, FILE_WRITE);
  copier.begin(file, out);  
  recording = true; 
}
```

and this is the record_end function:
```
void record_end(bool pinStatus, int pin, void* ref){
  if (recording == true){
    Serial.println("Playing...");
    file.close();
    recording = false;
    file = SD.open(file_name); // reopen in read mode
    copier.begin(kit, file);  // start playback
  }
}
```

My final goal is to make a recording from the mic, encode as WAV and save it to SD card and after that send the file over bluetooth to an Android smartphone. I am working on a recording device which will be used for collecting sounds for an audio classifier. The android app will let the users add tags to the recordings and upload them to our cloud storage. The project is with a great cause - saving bees by analysing their sounds!

**Comment by Nik-mayak** [comment]
I will answer you a little later :)


чт, 4 мая 2023 г., 10:26 Miroslav Petrov ***@***.***>:

**Comment by pschatzmann** [comment]
If you follow the Wiki about encoding and decoding the excercise should be trivial.
Anyhow I have added a simple [example using the AudioKit](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-audiokit-sd_wav/streams-audiokit-sd_wav.ino) 
As described in the Wiki, the critical thing is to close or flush the file in the end, to store the data! But this has nothing to do with my library...

If you want to process another output do not forget to call out.begin(info); again. Otherwise the output does not know that you restarted and it will not write a wav header...

**Reply by miroslavpetrov** [reply]
Thank you very much for the provided example! However, I cannot compile it...Here is the compiler output for reference:

```
/private/var/folders/cf/dxpq_74x0gv1884hrc2nhsf80000gn/T/.arduinoIDE-unsaved202344-11125-1djdqok.e9bs/sketch_may4a/sketch_may4a.ino:18:1: error: 'AudioInfo' does not name a type; did you mean 'AudioKitInfo'?
 AudioInfo info(16000, 1, 16);
 ^~~~~~~~~
 AudioKitInfo
/private/var/folders/cf/dxpq_74x0gv1884hrc2nhsf80000gn/T/.arduinoIDE-unsaved202344-11125-1djdqok.e9bs/sketch_may4a/sketch_may4a.ino: In function 'void setup()':
/private/var/folders/cf/dxpq_74x0gv1884hrc2nhsf80000gn/T/.arduinoIDE-unsaved202344-11125-1djdqok.e9bs/sketch_may4a/sketch_may4a.ino:42:16: error: 'info' was not declared in this scope
   cfg.copyFrom(info);
                ^~~~
/private/var/folders/cf/dxpq_74x0gv1884hrc2nhsf80000gn/T/.arduinoIDE-unsaved202344-11125-1djdqok.e9bs/sketch_may4a/sketch_may4a.ino:42:16: note: suggested alternative: 'in'
   cfg.copyFrom(info);
                ^~~~
                in


Compilation error: 'AudioInfo' does not name a type; did you mean 'AudioKitInfo'?
```

**Reply by pschatzmann** [reply]
Then you have still a very old version of my library:
Upgrade or replace the AudioInfo with individual fields

**Reply by miroslavpetrov** [reply]
I have updated the library and it now compiles. I have also saw that you made another commit to the example. There is a wav file on the sd-card that is not 0kb but the recording has no sound.

**Reply by pschatzmann** [reply]
Hmm, I did not really check if there is any sound, because when I saw that the file contains data, I assumed that everything is ok. After all it is just a simple copy from the source to the file, what could go wrong here?

Just make a test and I can reproduce your problem!

**Comment by Nik-mayak** [comment]
[main.zip](https://github.com/pschatzmann/arduino-audio-tools/files/11397839/main.zip)
[platformio.zip](https://github.com/pschatzmann/arduino-audio-tools/files/11397850/platformio.zip)
This is the working code with autoresponder algorithm. You can cut out the part of the code related to the entry.

**Comment by pschatzmann** [comment]
I just nailed down where the issue is: In the AudioKit the RX mode seens to be broken and does not produce any signal any more.
Did you try get a file w/o sound with I2SStream or AudioKitStream ? 
Opening it in RXTX_MODE seems to work...

**Reply by pschatzmann** [reply]
The issue has been introduced when I replaced the custom I2S implementation in the AudioKit with the regular I2SStream from my framework. I tested the TX_MODE and RXTX_MODE and assumed that the RX_MODE would work as well: it seems that the pins were not configured correctly for this case, therefore there was no signal.

The WAVE recording part of the example has always been working correctly!

I have committed the correction and tested the WAV file in Audacity...

**Reply by miroslavpetrov** [reply]
Now it works as expected! Thank you for your time and great work!
